### PR TITLE
Call TriggerGlobalGC when the plasma store is full

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

If the plasma store is full, it may be due to cyclic references to `ObjectID`s in python somewhere on the cluster. This change causes workers to trigger a global GC in an attempt to garbage collect such references and free up space before giving up when the plasma store is full.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by edoakes** [comment]
Dependent on https://github.com/ray-project/ray/pull/7328 and still need to add tests.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc [hunk 1]
```diff
@@ -55,28 +56,49 @@ Status CoreWorkerPlasmaStoreProvider::Create(const std::shared_ptr<Buffer> &meta
...
       message << "Failed to put object " << object_id << " in object store because it "
               << "is full. Object size is " << data_size << " bytes.";
-      return Status::ObjectStoreFull(message.str());
+      status = Status::ObjectStoreFull(message.str());
+      if (max_retries < 0 || retries < max_retries) {
+        RAY_LOG(ERROR) << message.str() << " Plasma store status:\n"
+                       << MemoryUsageString() << "\nWaiting " << delay
+                       << "ms for space to free up...";
+        if (on_store_full_) {
+          on_store_full_();
```
Comment (ericl) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc]: Should we call this once or multiple times? I guess it's possible you might need a series of gc calls, though I'm not sure about the scenario.

**Review by ericl — COMMENTED** [review]
Can we add an end to end Python test for this, similar to the one in the ref count tests?

**Comment by ericl** [comment]
Oh nevermind, I see it now.

**Review by ericl — APPROVED** [review]

File: src/ray/core_worker/store_provider/plasma_store_provider.cc (hunk 1 above)
Comment (edoakes) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Let's call it multiple times for now to cover all scenarios. This is throttled anyways to once per heartbeat interval across the cluster.

**Comment by ericl** [comment]
Here's another test you can add that introduces cyclic refs in both the worker and driver:
```
import ray
import gc
import random
import time
import json
import numpy as np


ray.init(num_cpus=1, object_store_memory=2000 * 1024 * 1024)

def block():
    return np.zeros(100 * 1024 * 1024, dtype=np.uint8)


class CyclicRef:
    def __init__(self, v):
        self.value = v
        self.cyclic_ref = self


refs = [
    CyclicRef(ray.put(block())) for _ in range(10)
]


@ray.remote
def f(x):
    y = CyclicRef(x)
#    gc.collect()
    return x


for j in range(1000):
    i = random.randint(0, len(refs) - 1)
    res = f.remote(refs[i].value)
    refs[i] = CyclicRef(res)
    ray.get(res)
    print("processing", j)
#    gc.collect()
```

Couple other suggestions:
- only print "gc collect freed ...." in raylet.pyx if num freed refs > 0, to avoid spam
- only print the object store memory message if the all retries failed

**Pull request merged**